2단계 미션 제출합니다.

minleejae opened the pull request:
## 변경사항

1. 회수 입력에 대한 검증 과정을 추가했습니다.
2. MVC 패턴을 적용하여 리팩토링하였습니다.
3. 랜덤 숫자를 사용하는 메서드를 테스트할 수 있도록 랜덤 숫자 생성 과정을 변경하고 테스트를 추가했습니다.

minleejae committed:
feat: 회수 입력 검증 기능 추가

minleejae committed:
refactor: 메서드 순서 변경

minleejae committed:
refactor: RacingGame MVC 구조 적용

minleejae committed:
refactor: 랜덤 숫자를 이용하는 메서드를 테스트 가능하도록 변경

랜덤 숫자를 생성하는 기능을 strategy 패턴을 이용하도록 변경하고,
테스트를 추가했습니다.

seokhong commented on src/main/java/racingcar/view/RacingGameView.java:
```diff
@@ -25,20 +32,20 @@ public List<String> requestCarNames() {
-        return Integer.parseInt(sc.nextLine());
+        int numberOfTrials = Integer.parseInt(sc.nextLine());
+
+        if (numberOfTrials <= 0) {
```
진행 횟수 검증을 뷰에서 하도록 구현하셨네요!
뷰와 도메인의 밸리데이션은 어떤 차이가 있을까요? `뷰 도메인 밸리데이션` 과 같은 키워드로 찾아보시면 도움이 될 것 같습니다.

minleejae replied:
뷰와 도메인 밸리데이션을 구분하고, 도메인 밸리데이션을 추가했습니다. [47ce5cf](https://github.com/next-step/java-racingcar-kakao/pull/44/commits/47ce5cf789a1993721c07efe43f4ac2ace9eb935)

seokhong commented on src/main/java/racingcar/controller/RacingGameController.java:
```diff
@@ -0,0 +1,52 @@
+    public void playGame() {
+        int trial = requestTrial();
+
+        RacingGame racingGame = initializeRacingGame();
```
진행 횟수도 `RacingGame`이 직접 관리해보는 건 어떨까요?
그럼 21행의 for 문에서 진행 횟수에 대한 제어를 `RacingGameController`가 하는 대신, `RacingGame에게 게임이 종료됐는지 묻는 형태`로도 구현이 가능할 것 같아요!

minleejae replied:
RacingGame이 진행 횟수를 관리하도록 변경했습니다. [7a250cd](https://github.com/next-step/java-racingcar-kakao/pull/44/commits/7a250cd3faeaf803199ee3b2c7d459b40341cb7f)

seokhong commented on src/main/java/racingcar/generator/RandomNumberGenerator.java:
```diff
@@ -0,0 +1,13 @@
+
+    @Override
+    public int generate() {
+        return random.nextInt(10);
```
10이라는 값도 상수로 관리해볼 수 있겠네요. :)

minleejae replied:
상수로 관리하도록 변경했습니다. [6980b3f](https://github.com/next-step/java-racingcar-kakao/pull/44/commits/6980b3f6f1352adc5fe14f95c5626d5a68045ce4)

seokhong commented on src/main/java/racingcar/controller/RacingGameController.java:
```diff
@@ -0,0 +1,52 @@
+
+    private RacingGame initializeRacingGame() {
+        List<String> carNames = requestCarNames();
+        racingGameView.startGameRound();
```
이 문구의 출력은 게임을 진행하기 전에 해야 하지 않을까요?
`RacingGame`을 초기화하는 과정에서 문구를 출력해야 할 필요가 있을까요?

minleejae replied:
문구 출력 시점을 변경했습니다. [5c3a501](https://github.com/next-step/java-racingcar-kakao/pull/44/commits/5c3a501cce1d0759dcfa3e665beaa6b6d3bb638b)

seokhong commented on src/main/java/racingcar/model/RacingGame.java:
```diff
@@ -0,0 +1,41 @@
+                .filter(car -> car.isAt(maxMoves))
+                .collect(Collectors.toList());
+    }
+}
```
파일 마지막에 newline 이 없다고 깃헙에서 경고를 띄워주네요.  
파일 마지막의 newline 은 무슨 역할을 하는지 학습해보신 후, IDE 설정을 통해 휴먼 에러를 방지해보시면 좋겠습니다.

minleejae replied:
파일 마지막에 newline을 추가했습니다. [6980b3f](https://github.com/next-step/java-racingcar-kakao/pull/44/commits/6980b3f6f1352adc5fe14f95c5626d5a68045ce4)

seokhong commented on src/main/java/racingcar/model/RacingGame.java:
```diff
@@ -0,0 +1,41 @@
+        racingCars.forEach(RacingCar::move);
+    }
+
+    public List<RacingCar> getRacingCars() {
```
이 메서드가 반환한 값을 외부에서 조작하면 `RacingGame`이 관리하던 `racingCars`에도 영향이 갈 것으로 보이네요.
방어적 복사 등의 키워드를 찾아보신 후 적용해보는 건 어떨까요?

minleejae replied:
방어적 복사를 적용했습니다. [7a250cd](https://github.com/next-step/java-racingcar-kakao/pull/44/commits/7a250cd3faeaf803199ee3b2c7d459b40341cb7f)

seokhong commented on src/test/java/racingcar/RacingCarTest.java:
```diff
@@ -35,29 +40,42 @@ void stopTest(int number) {
-        racingCars.getRacingCars().get(0).getNumAndMove(9);
-        racingCars.getRacingCars().get(1).getNumAndMove(2);
-        racingCars.getRacingCars().get(2).getNumAndMove(8);
+        racingGame.getRacingCars().get(0).getNumAndMove(9);
```
`testPerformRacingRound`를 보면 외부에서 값을 제어해서 이동을 제어하는 방법을 찾으신 것으로 보이네요!
아마 우승자를 찾기 위해서는 자동차마다 이동 여부를 다르게 해야하고, 현재 `TestNumberGenerator`의 구현 상 매번 다른 값을 반환하는 형태가 아니라서 객체를 꺼내서 이동시키신 것으로 보입니다. `TestNumberGenerator`의 구현을 수정해서, 생성하고 싶은 값을 리스트로 받아보는 건 어떨까요? 그럼 car1, car2, car3를 위한 값을 생성할 때 9, 2, 8을 순서대로 생성하도록 제어하는 것도 가능할 것으로 보이네요.

minleejae replied:
TestNumberGenerator를 배열을 바탕으로 수를 생성하도록 변경했습니다.[0a2ed5f](https://github.com/next-step/java-racingcar-kakao/pull/44/commits/0a2ed5f0f9624d28d21a483e68b37444c894c50c)

seokhong reviewed (changes requested):
민재님, 2단계 미션 잘 진행하셨습니다. 👍 
1단계 미션에 남긴 코멘트 반영하신 것 확인했습니다.
2단계 미션에서 조금 더 고민해보시면 좋을 것 같은 내용을 코멘트로 남겼습니다. 코멘트 확인 부탁드립니다. 🙂

minleejae committed:
refactor: 상수 관리, 마지막 newline 추가

minleejae committed:
refactor: 실행 결과 출력 메서드 호출 시점 변경

minleejae committed:
refactor: RacingGame 진행 횟수 관리, 방어적 복사 적용

minleejae committed:
refactor: RacingGame 생성자 중복 제거

minleejae committed:
refactor: RacingGame 사용자 입력 검증 역할 수정

기존에는 컨트롤러에서 에러를 처리하도록 했는데,
입력에 관한 에러는 뷰에서 바로 처리하도록 변경했습니다.

뷰와 도메인에 관한 밸리데이션을 구분하기 위해
뷰에서는 입력 값이 숫자가 맞는지 검증하도록 수정했습니다.

minleejae committed:
feat: RacingGame 도메인 밸리데이션 추가

minleejae committed:
refactor: TestNumberGenerator 배열 바탕으로 생성

seokhong commented on src/test/java/racingcar/TestNumberGenerator.java:
```diff
@@ -0,0 +1,19 @@
+
+    @Override
+    public int generate() {
+        return numbersIterator.hasNext() ? numbersIterator.next() : 0;
```
`NoSuchElementException`이 발생하지 않도록 꼼꼼한 체크 좋습니다. 👍

seokhong commented on src/test/java/racingcar/RacingCarTest.java:
```diff
@@ -0,0 +1,84 @@
+
+    @ParameterizedTest
+    @ValueSource(ints = {4, 5, 6, 7, 8, 9})
+    void moveTest(int number) {
```
`isMove`, `getNumAndMove` 메서드는 테스트를 제외하면 프로덕션 코드에서는 `RacingCar` 내부에서만 사용될 것으로 보이네요.
테스트를 위해 접근 제어자를 public 으로 변경하는 것은 지양하는 것이 좋습니다. 테스트 코드는 public 메서드가 의도한대로 동작하는지 확인하기 위함이지, 구체적인 구현을 검증하기 위한 것이 아닙니다. private 메서드를 직접 검증하는 것은 구체적인 구현을 모두 알고 검증하는 것처럼 보입니다.
private 메서드에 대한 테스트는 public 메서드를 검증하는 과정에서 함께 검증되어야 합니다. `RacingCar`를 생성할 때 `TestNumberGenerator`를 전달하고 난수 생성을 제어할 수 있다면, `move` 메서드를 테스트하면서 `isMove`에 대한 테스트도 될 것입니다. `move` 메서드만 테스트하는 형태로 작성함으로써, `move`의 세부 사항이 달라지더라도 테스트는 변경되지 않을 수 있습니다.

minleejae replied:
테스트를 위해 public으로 선언했던 메서드들을 private으로 변경 후 ```move```를 통해 테스트하도록 코드를 변경하였습니다. [b46bb31](https://github.com/next-step/java-racingcar-kakao/pull/44/commits/b46bb31a8fcefd68d74cd8cbc83b415cf3df8c5e)

seokhong commented on src/test/java/racingcar/RacingCarTest.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    @ParameterizedTest
+    @ValueSource(ints = {4, 5, 6, 7, 8, 9})
```
모든 값을 테스트하는 것도 좋지만, 값의 개수가 많아질수록 테스트 작성에 어려움이 있을 수 있습니다.
경계값 테스트 라는 개념을 찾아보신 후 적용해보셔도 좋겠습니다.

minleejae replied:
경계값을 통해 테스트를 하도록 코드를 수정했습니다 [b46bb31](https://github.com/next-step/java-racingcar-kakao/pull/44/commits/b46bb31a8fcefd68d74cd8cbc83b415cf3df8c5e)

seokhong commented on src/test/java/racingcar/RacingCarTest.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    @Test
+    void determineWinnerTest() {
```
`determineWinnerTest`, `testPerformRacingRound`는 `racingGame`에 대한 테스트이므로 `RacingGameTest` 파일로 이동하여 관리해보시면 좋겠습니다. :)

minleejae replied:
테스트 코드를 분리하였습니다. [1ec5092](https://github.com/next-step/java-racingcar-kakao/pull/44/commits/1ec5092d51005ac22df7d0bfb7a41f5fd51fc0c1)

seokhong commented on src/main/java/racingcar/model/RacingCar.java:
```diff
@@ -1,30 +1,37 @@
     }
-
+    
     public void getNumAndMove(int num) {
```
이전 PR의 [코멘트](https://github.com/next-step/java-racingcar-kakao/pull/21#discussion_r1541149581)에 따르면, 중요 로직 순서(public 메서드 중 자주 사용하는 것이 상단에 온다는 의미일까요? 또는 도메인에서 중요하다고 생각하시는 메서드가 상단에 온다는 의미일까요?) 또는 public 메서드에 이어 내부 구현되는 private 메서드가 배치되어야 할 것으로 보이네요.
현재의 도메인 클래스에 구현된 메서드의 순서를 보면 getter 가 중간에 있는 것을 포함하여, 다소 섞여 있는 것으로 보입니다. 대개 getter, setter, toString, equals, hashCode 등의 메서드는 하단에 배치합니다.

minleejae replied:
이전에는 public 메서드를 위로 보내는 것으로 생각했었는데, 말씀해주신 것 처럼 이어서 구현되는 메서드들을 순서대로 작성하는 게 좋아보이네요!
getter와 같은 메서드를 맨 아래에 위치하도록 하고, 메서드 순서를 읽기 쉽도록 재배치하였습니다.  [c9b6aae](https://github.com/next-step/java-racingcar-kakao/pull/44/commits/c9b6aae759591eedc3816404e708be052810674a)

seokhong reviewed (changes requested):
민재님, 2단계 미션 코멘트 잘 반영하셨습니다. 👍 
조금 더 고민해보시면 좋을 것 같은 내용 코멘트로 남겼습니다.
코멘트 확인 부탁드립니다. 🙂

minleejae committed:
refactor: RacingCar 접근제어자 변경, move 테스트 변경

테스트만을 위해 public으로 선언한 메서드를 private로 변경하고 해당 테스트 과정을 수정했습니다.

minleejae committed:
refactor: RacingCarTest 테스트 코드 분리

minleejae committed:
refactor: 메서드 순서 재배치

seokhong commented on src/main/java/racingcar/model/RacingCar.java:
```diff
@@ -1,32 +1,24 @@
     public RacingCar(String name, NumberGenerator numberGenerator) {
         validateName(name);
         this.name = name;
         this.numberGenerator = numberGenerator;
```
사소한 부분이지만 `numberGenerator`에 대한 null check 도 추가하면 더 안정적인 프로그램이 되겠네요. :)

minleejae replied:
꼼꼼한 피드백 감사합니다!

seokhong reviewed (approved):
민재님, 2단계 미션 코멘트 잘 반영하셨습니다. 💯 
2단계 PR 머지하겠습니다.

첫 미션 진행하는 동안 수고 많으셨습니다.
이번 주도 즐거운 한주 보내세요. 🙁
